### PR TITLE
Updated ext-email support

### PR DESCRIPTION
**Pull request by erenatas**
Added also a test case for version specific changes

Signed-off-by: Eren ATAS <eatas.contractor@libertyglobal.com>

### CODE REVIEW

File: jenkins_jobs/modules/publishers.py [hunk 1]
```diff
@@ -2456,37 +2456,119 @@ def base_email_ext(registry, xml_parent, data, ttype):
...
+            XML.SubElement(email, "includeCulprits").text = "false"
+            XML.SubElement(email, "sendToRecipientList").text = "true"
     if ttype == "ScriptTrigger":
         XML.SubElement(trigger, "triggerScript").text = data["trigger-script"]
 
+    if plugin_version >= pkg_resources.parse_version("2.39"):
+        mappings = [
+            ("attachments", "attachmentsPattern", ""),
+            ("attach-build-log", "attachBuildLog", False),
+            ("compress-log", "compressBuildLog", False),
```
Comment (romanek-adam) [inline jenkins_jobs/modules/publishers.py]: I guess these new API keys should be documented.

File: jenkins_jobs/modules/publishers.py [hunk 2]
```diff
@@ -2456,37 +2456,119 @@ def base_email_ext(registry, xml_parent, data, ttype):
...
         if "upstream-committers" in data["send-to"]:
             recipient_providers = XML.SubElement(email, "recipientProviders")
             XML.SubElement(
                 recipient_providers,
                 "hudson.plugins.emailext.plugins.recipients.UpstreamComitterRecipientProvider",
             ).text = ""
+        if plugin_version >= pkg_resources.parse_version("2.39"):
+            for recipient in data["send-to"]:
+                if "developers" == recipient:
+                    if "recipient_providers" not in locals():
```
Comment (romanek-adam) [inline jenkins_jobs/modules/publishers.py]: I'd rather set `recipient_providers` to `None` at the beginning of the whole code and check the variable for `None` later on. Using `locals()` is for such cases is rather a bad practice.

File: jenkins_jobs/modules/publishers.py (hunk 1 above)
Comment (erenatas) [reply jenkins_jobs/modules/publishers.py]: They are not changeable values anymore, therefore I don't think we need to document them.

File: jenkins_jobs/modules/publishers.py [hunk 3]
```diff
@@ -2456,37 +2456,117 @@ def base_email_ext(registry, xml_parent, data, ttype):
...
+            XML.SubElement(email, "sendToRequester").text = str(
+                "requester" in data["send-to"]
+            ).lower()
+            XML.SubElement(email, "includeCulprits").text = str(
+                "culprits" in data["send-to"]
+            ).lower()
+            XML.SubElement(email, "sendToRecipientList").text = str(
+                "recipients" in data["send-to"]
+            ).lower()
         if "upstream-committers" in data["send-to"]:
```
Comment (romanek-adam) [inline jenkins_jobs/modules/publishers.py]: Hint (minor): according to the documentation for JJB API this value should also be put under >= 2.39 condition below

File: jenkins_jobs/modules/publishers.py [hunk 4]
```diff
@@ -2456,37 +2456,117 @@ def base_email_ext(registry, xml_parent, data, ttype):
...
+            # in order to match the XML generated via UI.
+            failure_count_supporters = ["FirstFailureTrigger", "SecondFailureTrigger"]
+            if ttype in failure_count_supporters:
+                XML.SubElement(trigger, "failureCount").text = "0"
     else:
-        XML.SubElement(email, "sendToRequester").text = "false"
-        XML.SubElement(email, "sendToDevelopers").text = "false"
-        XML.SubElement(email, "includeCulprits").text = "false"
-        XML.SubElement(email, "sendToRecipientList").text = "true"
+        if plugin_version < pkg_resources.parse_version("2.39"):
```
Comment (romanek-adam) [inline jenkins_jobs/modules/publishers.py:2563]: Don't you need to handle the else case? I.e. no 'send-to' in `data` and version >= 2.39?

According to JJB docs recipients is enabled in 'send-to' by default, so I'm curious if this should be handled somehow.

File: tests/publishers/fixtures/email-ext004.yaml [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+      matrix-trigger: only-configurations
+      presend-script: "cancel=true"
+      postsend-script: "cancel=true"
+      save-output: true
+      send-to:
+        - developers
+        - requester
+        - culprits
+        - recipients
+        - upstream-committers
```
Comment (romanek-adam) [inline tests/publishers/fixtures/email-ext004.yaml:37]: It's worth adding:
* failing-test-suspects-recipients
* first-failing-build-suspects-recipients

File: jenkins_jobs/modules/publishers.py (hunk 4 above)
Comment (erenatas) [reply jenkins_jobs/modules/publishers.py:2563]: It's already handled above. The naming conventions have changed after the version

File: jenkins_jobs/modules/publishers.py (hunk 3 above)
Comment (romanek-adam) [reply jenkins_jobs/modules/publishers.py]: As discussed please make it work only for 2.39 and above.

File: jenkins_jobs/modules/publishers.py (hunk 4 above)
Comment (romanek-adam) [reply jenkins_jobs/modules/publishers.py:2563]: But the above code block is only executed if `send-to` key is in `data`.

I think when `send-to` IS NOT in `data` and version >= 2.39 then JJB defaults are not honored (i.e. `ListRecipientProvider` should be added to XML as `recipients` is marked as true/enabled by default, and that's why I think you need to handle the else case).

What do you think?

File: jenkins_jobs/modules/publishers.py [hunk 6]
```diff
@@ -2462,37 +2462,124 @@ def base_email_ext(registry, xml_parent, data, ttype):
...
+            XML.SubElement(email, "sendToRequester").text = str(
+                "requester" in data["send-to"]
+            ).lower()
+            XML.SubElement(email, "includeCulprits").text = str(
+                "culprits" in data["send-to"]
+            ).lower()
+            XML.SubElement(email, "sendToRecipientList").text = str(
+                "recipients" in data["send-to"]
+            ).lower()
+        if plugin_version >= pkg_resources.parse_version("2.39"):
```
Comment (romanek-adam) [inline jenkins_jobs/modules/publishers.py]: This should be `else:` to the above `if` section.

File: jenkins_jobs/modules/publishers.py [hunk 7]
```diff
@@ -2462,37 +2462,124 @@ def base_email_ext(registry, xml_parent, data, ttype):
...
+                    recipient_providers,
+                    "hudson.plugins.emailext.plugins.recipients.UpstreamComitterRecipientProvider",
+                ).text = ""
+    else:
+        if plugin_version < pkg_resources.parse_version("2.39"):
+            XML.SubElement(email, "sendToRequester").text = "false"
+            XML.SubElement(email, "sendToDevelopers").text = "false"
+            XML.SubElement(email, "includeCulprits").text = "false"
+            XML.SubElement(email, "sendToRecipientList").text = "true"
+        elif plugin_version >= pkg_resources.parse_version("2.39"):
```
Comment (romanek-adam) [inline jenkins_jobs/modules/publishers.py]: Again, this should be a simple `else` to the above `if` section.

**Comment by romanek-adam** [comment]
@erenatas at this point we should consider upstreaming to see what JJB maintainers would say about these changes.

**Comment by erenatas** [comment]
> @erenatas at this point we should consider upstreaming to see what JJB maintainers would say about these changes.

https://review.opendev.org/#/c/749071/1

**Comment by romanek-adam** [comment]
> > @erenatas at this point we should consider upstreaming to see what JJB maintainers would say about these changes.
> 
> https://review.opendev.org/#/c/749071/1

Please update the change with the latest changes to this PR and keep them aligned.

**Comment by erenatas** [comment]
> > > @erenatas at this point we should consider upstreaming to see what JJB maintainers would say about these changes.
> > 
> > 
> > https://review.opendev.org/#/c/749071/1
> 
> Please update the change with the latest changes to this PR and keep them aligned.

Looks good now :)

**Review by romanek-adam — APPROVED** [review]

**Pull request merged**